netstandard.md

```diff
@@ -0,0 +1,127 @@
+# Dropping support for .NET Standard
+
+## Background
+
+The OpenTelemetry .NET project depends heavily on APIs provided by .NET.
+Specifically, the APIs provided under `System.Diagnostics.DiagnosticSource`.
+
+With the release of .NET 6, [it was announced](https://github.com/dotnet/announcements/issues/190)
+that the .NET team will drop out-of-support frameworks for a number of packages
+including `System.Diagnostics.DiagnosticSource`. This practice will continue
+with the release of .NET 7. Frameworks no longer supported will include .NET
+Core 3.1 and .NET 5. Refer to the [.NET download](https://dotnet.microsoft.com/download/dotnet)
+page to view the end of support dates for each version of .NET.
+
+The core packages offered by OpenTelemetry .NET currently ship a .NET
+Standard build (i.e., `netstandard2.0` and/or `netstandard2.1`).
+Therefore, OpenTelemetry .NET can technically be consumed by projects targeting
+out-of-support frameworks - even as far back as .NET Core 2.0.
+
+However, referencing the latest version of OpenTelemetry .NET by an application
+targeting .NET Core 3.1 or .NET 5, for example, will generate the following
+build warnings:
+
+> System.Diagnostics.DiagnosticSource doesn't support netcoreapp3.1. Consider updating your TargetFramework to net6.0 or later.
+
+> System.Diagnostics.DiagnosticSource doesn't support net5.0. Consider updating your TargetFramework to net6.0 or later.
+
+These warnings can be suppressed by setting the
+`SuppressTfmSupportBuildWarnings` MSBuild property. However,
+OpenTelemetry .NET is not tested against out-of-support frameworks. Therefore
+there is no guarantee that it will continue to function.
+
+## The plan to remove .NET Standard targets
+
+The plan moving forward is to align the support cycle of OpenTelemetry .NET
+with the support cycle of .NET. To achieve this, OpenTelemetry .NET plans to
+drop .NET Standard builds of its artifacts. OpenTelemetry packages will only
+offer implementation specific target frameworks (e.g., `net462` and `net6.0`).
+This will enable us to easily test and validate that OpenTelemetry .NET works
+as expected for all the frameworks we target.
+
+At this time, the frameworks targeted will include at least `net462` and
+`net6.0`, but may include others like `net7.0`, when appropriate. As target
+frameworks reach end of life, we will remove those targets during a major
```

Comment:
major version is something like 2.0 / 3.0?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/2138 historically we only did minor version bump when we dropped framework support (when the fw itself went out-of-support)

Reply:
Yes, this differs from our prior pattern of practice, and I'm suggesting we reconsider our practice for the future. As mentioned in https://github.com/dotnet/announcements/issues/190

> Dropping a framework from a package is a source breaking change.

I think a major version number is the correct tool to communicate this to users clearly. It aligns with the .NET Runtime team's versioning of DiagnosticSource in regards to dropping supported frameworks.

That said, there certainly are other examples of library authors dropping frameworks with minor version releases of their libraries. Though, I do not think a minor version bump communicates very clearly whether the drop in support was intentional or not.

I'd be very interested in some official guidance to library authors in this respect. Maybe here https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/versioning. Something like:

> ✔️ CONSIDER (or DO?) bumping the major version when dropping framework support.

Maybe .NET would consider this overstepping their bound with their guidance, but I'd like to hear their thoughts.

Reply:
My main worry with major version bump was the above _( worried about upgrading from 2.0 to 3.0 due to the fear of breaking changes)_

I am more inclined to do minor version bump, and document that some runtimes are no longer supported, as opposed to doing major version bump and documenting no-breaking-api-but-drop-certain-runtimes)

Reply:
sorry, didn't see @reyang's comment prior to posting mine.

> For example, a user might be worried about upgrading from 2.0 to 3.0 due to the fear of breaking changes, although what actually happened is just a drop of deprecated runtime version.

Would we anticipate that users upgrading to OTel 1.4 would be worried that it comes with a major version bump of DiagnosticSource 6 -> 7. That is, might they be concerned that 7 introduces breaking changes?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/1769
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2179

We make a bet than System.* and Ms.Extensions.Logging.* will not have breaking API changes *even with major version bump*!
This is one of the reasons library authors should feel more comfortable taking a dependency on DS (as opposed to OTel.API which can have breaking changes with Major ver bump).

Reply:
To summarize, I think we have the following options:

1. Bump minor version when we drop old runtime versions. And communicate it clearly (via docs or something else).
2. Bump major version when we drop old runtime versions, and communicate it clearly that OpenTelemetry .NET will never make intentional breaks in it's public API surface and data semantics. Which essentially means https://github.com/open-telemetry/opentelemetry-dotnet/labels/need-majorversion-bump will have to be solved in a backward-compatible way rather than a breaking way.
3. Bump major version when we drop old runtime versions, and have clear communication whether a major version bump was because of pure runtime version drop, or API surface break, or a combination of both.
4. Bump major version whenever we drop old runtime or break API surface, without communicating whether the verion bump was a result of runtime version deprecation or API breaks.

I'm fine with either 1, 2 or 3. I would rank them 1 > 2 > 3.
I think we should avoid 4.

Reply:
> it will need to change the current position: a major version bump is NOT going to introduce breaks to the API surface or any telemetry data semantics

Yes, fair. Personally, I think this would be a worthy goal in the long run. So basically adopting option number 2.

In the short run though, I was more considering option number 3. Version 1.x of OTel .NET has served as a learning grounds as OpenTelemetry has matured. We already have a number of things marked `Obsolete` and we're anticipating some more. I think cleaning these things up in a breaking major release will end up benefitting users in the long run.

---

Though let me ask a different question. Do we all agree we want to drop .NET Standard targets from our core packages in the future? If so, my primary intent with this announcement is to:
1) Give end users a very clear notice of our intended plan to drop .NET Standard, and
2) Some sense of a rough timeline for when we will do this.

Whether this be in a minor or major release is more of a technical detail.

Reply:
In other words, I'm proposing this announcement as our resolution to #3448.

Reply:
#3703 as an alternative resolution to #3448.

Reply:
> To summarize, I think we have the following options: ...

IMO, option (3) is the best one. I think many people assume minor version number bump = "not a breaking change, may add features, no need to even look at release notes if I don't feel like it." Doing a minor version bump that broke people because it dropped support for the runtime they were using would be unexpected.

For option (2) I'd be careful about promises that you will never make a breaking change, it could become a real straight jacket. I'd suggest something more like you strongly value compatibility and breaking changes should be rare.

Reply:
Thank you for your guidance @noahfalk!

> Doing a minor version bump that broke people because it dropped support for the runtime they were using would be unexpected.

This is the primary reason I'm suggesting changing our practice to a major version bump under this proposal is because with the absence of .NET Standard targets, dropping support for a runtime version means users on old frameworks can not upgrade.

Previously, when we dropped `net461`, `netcoreapp3.1`, and `net5.0` targets we have done it with a minor version bump. Though, in all of these instances, users targeting these frameworks can currently still upgrade as they will fall back to our `netstandard2.x` targets. It is for this reason I was not as inclined to recommend a major version bump.

@cijothomas @CodeBlanch @utpilla Maybe we chat once more in next week's SIG meeting before choosing between this proposal or #3703? If we're set on dropping .NET Standard, we can also further discuss the topic of major vs. minor version bump.